### ISSUE TITLE
Git regression from npm@2.1.17 onward: npm can't install git+ssh dependencies anymore

### ISSUE THREAD

**Issue by madarche** [npm/npm#8031]
Hello,

From npm@2.1.17 onward (I've tested all the versions from 2.1.17 to 2.8.4) npm can't install git+ssh dependencies.

Here are examples of used dependencies:

    "dependencies": {
      "secret-factory": "git+ssh://git@git.example.test:secret-factory.git#v1.0.0",
      "sso-connector": "git+ssh://git@git.example.test:sso-connector-js.git#v1.0.0"
    },
    "devDependencies": {
      "sso-mockup": "git+ssh://git@git.example.test:sso-mockup.git#v3.0.0"
    },

The presence or not of an `npm-shrinkwrap.json` has no effect.

Cheers

**Comment by madarche** [comment]
Here is the log for npm@2.8.4:

    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7: Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7'...
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7: fatal: '/sso-connector-js.git' does not appear to be a git repository
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7: fatal: Could not read from remote repository.
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7:
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7: Please make sure you have the correct access rights
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-connector-js.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7: and the repository exists.
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80: Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80'...
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80: fatal: '/sso-mockup.git' does not appear to be a git repository
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80: fatal: Could not read from remote repository.
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80:
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80: Please make sure you have the correct access rights
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/sso-mockup.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80: and the repository exists.
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639: Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639'...
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639: fatal: '/secret-factory.git' does not appear to be a git repository
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639: fatal: Could not read from remote repository.
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639:
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639: Please make sure you have the correct access rights
    npm ERR! git clone --template=/home/sweet/.npm/_git-remotes/_templates --mirror ssh://git@git.example.test/secret-factory.git /home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639: and the repository exists.
    npm ERR! Linux 3.2.0-4-amd64
    npm ERR! argv "node" "/home/sweet/.nvm/v0.10.33/bin/npm" "install"
    npm ERR! node v0.10.33
    npm ERR! npm  v2.8.4
    npm ERR! code 128

    npm ERR! Command failed: Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7'...
    npm ERR! fatal: '/sso-connector-js.git' does not appear to be a git repository
    npm ERR! fatal: Could not read from remote repository.
    npm ERR!
    npm ERR! Please make sure you have the correct access rights
    npm ERR! and the repository exists.

**Comment by madarche** [comment]
Here is the log for npm@2.1.17:

    npm ERR! git clone ssh://git@git.example.test/secret-factory.git Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-secret-factory-git-17bd1639'...
    npm ERR! git clone ssh://git@git.example.test/secret-factory.git fatal: '/secret-factory.git' does not appear to be a git repository
    npm ERR! git clone ssh://git@git.example.test/secret-factory.git fatal: Could not read from remote repository.
    npm ERR! git clone ssh://git@git.example.test/secret-factory.git
    npm ERR! git clone ssh://git@git.example.test/secret-factory.git Please make sure you have the correct access rights
    npm ERR! git clone ssh://git@git.example.test/secret-factory.git and the repository exists.
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-connector-js-git-93cc66e7'...
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git fatal: '/sso-connector-js.git' does not appear to be a git repository
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git fatal: Could not read from remote repository.
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git Please make sure you have the correct access rights
    npm ERR! git clone ssh://git@git.example.test/sso-connector-js.git and the repository exists.
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example-test-sso-mockup-git-5c2c7c80'...
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git fatal: '/sso-mockup.git' does not appear to be a git repository
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git fatal: Could not read from remote repository.
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git Please make sure you have the correct access rights
    npm ERR! git clone ssh://git@git.example.test/sso-mockup.git and the repository exists.
    npm ERR! Linux 3.2.0-4-amd64
    npm ERR! argv "node" "/home/sweet/.nvm/v0.10.33/bin/npm" "install"
    npm ERR! node v0.10.33
    npm ERR! npm  v2.1.17
    npm ERR! code 128

    npm ERR! Command failed: Cloning into bare repository '/home/sweet/.npm/_git-remotes/ssh-git-git-example.test-secret-factory-git-17bd1639'...
    npm ERR! fatal: '/secret-factory.git' does not appear to be a git repository
    npm ERR! fatal: Could not read from remote repository.
    npm ERR!
    npm ERR! Please make sure you have the correct access rights
    npm ERR! and the repository exists.

**Comment by yoni-tock** [comment]
Is this just on ubuntu? I'm seeing this on our Jenkins box (npm 2.7.4, node 0.12.2) but not locally on my OS X machine (npm 2.5.4 node 0.12.0).

**Comment by yoni-tock** [comment]
Looks like it may be a node version issue. I downgraded my Jenkins box to node 0.12.0 and it works again.

**Comment by thefourtheye** [comment]
Try if `git+https` works for you

**Comment by othiym23** [comment]
I think the issue is probably how your ssh URLs are formatted, but I need to do some testing before documenting it. I'll get to this today, most likely.

**Comment by madarche** [comment]
@yoni-tock I should have written that it's being experienced on different Unices (Debian and OS X).

@thefourtheye we really need git+ssh since the mentioned Git repositories are only accessible through SSH.

@othiym23 thanks a lot for your fast reply and concern. NPM is a wonderful tool! We use the `git+ssh://user@hostname:project.git#commit-ish` form documented in the [Git URLs as Dependencies](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) doc. The documented form `git+ssh://user@hostname/project.git#commit-ish` has never ever work for us, no matter the NPM version, no matter the client OS (Debian and OS X) and no matter the `git` version on the client.

**Comment by madarche** [comment]
The version of `git` on the server serving the NPM modules in question might be relevant too.

So here it is: it's `git version 1.7.2.5` running on a Debian GNU-Linux.

And I have just tested on another more current system (with a `git version 1.8.3.1` on a CentOS 7 and a dummy repository with URL `somerepo.git`) and it's the same behavior: the `git+ssh://user@hostname/project.git#commit-ish` form (with the slash instead of the colon) isn't accepted:

    fatal: repository 'somerepo.git' does not exist

May I ask for which systems/architectures is the `git+ssh://user@hostname/project.git#commit-ish` form working?

**Comment by othiym23** [comment]
I don't think this is a git version issue, but tell me: does this issue persist if you change the urls to be of the form `git+ssh:user@hostname/project.git#commit-ish` (without the two leading slashes)? Even if that fixes it, there's still an issue here where the documentation needs to be clearer about the two different kinds of git SSH URLs.

Sorry that I didn't get to this this week; I'm setting aside a week or two (more) to do nothing but deal with these git issues until we've got them all worked out.

**Comment by madarche** [comment]
Thanks @othiym23 for the suggestion! Here it the result: the form `git+ssh:user@hostname/project.git#commit-ish` (without the two leading slashes) doesn't work (tested with npm@1.4.28, npm@2.5.1 and npm@2.9.0) with the following error messages:

    ssh: Could not resolve hostname ssh: Name or service not known
    fatal: Could not read from remote repository
    Please make sure you have the correct access rights and the repository exists

**Comment by Zenwolf** [comment]
I have the same problem where I can use CLI npm install to install the private git+ssh correctly, but when it is included in the package.json and I use npm install it cannot find the repo anymore.

*Works:*
```
npm install --save git+ssh://git@github.com:Zenwolf/private-repo.git#branch-foo
``
*Fails afterwards:*
```
npm install

npm ERR! Darwin 12.6.0
npm ERR! argv "node" "/Users/mjaquish/local/bin/npm" "install"
npm ERR! node v0.10.26
npm ERR! npm  v2.8.3

npm ERR! version not found: mymodule@0.0.1
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/mjaquish/git/foo/npm-debug.log

**Comment by madarche** [comment]
Just to show this ticket is actively watched :-) : the git regression is still present in npm@2.9.1.

**Comment by milaniliev** [comment]
I've hit the exact same issue. It seems to me that npm is treating the presence of double slashes as indication to replace the ssh-colon with a slash.

**Comment by milaniliev** [comment]
Also, this does not seem to be a node or git problem: I tried it with node@0.10.38 and git@2.1.2, and nothing fixed it until I downgraded npm to 2.1.16. (Thanks @madarche!)

**Comment by madarche** [comment]
The git regression is still present in npm@2.10.0

**Comment by madarche** [comment]
The git regression is still present in npm@2.10.1

**Comment by thefourtheye** [comment]
I am not able to reproduce the problem.

    ➜  Desktop  /usr/local/bin/npm --version
    2.9.0
    ➜  Desktop  rm -rf node_modules/underscore/
    ➜  Desktop  /usr/local/bin/npm install  
    npm WARN package.json @ No description
    npm WARN package.json @ No repository field.
    npm WARN package.json @ No README data
    underscore@1.8.3 node_modules/underscore
    ➜  Desktop  cat package.json 
    {
        "dependencies": {
            "underscore_": "git+ssh://git@github.com:jashkenas/underscore.git"
        }
    }%

What am I missing?

**Comment by milaniliev** [comment]
@thefourtheye I don't know why that would be the case, but could you maybe try a host other than GitHub? I'm having this issue with my locally hosted packages.

**Comment by thefourtheye** [comment]
@milaniliev Sorry I didn't understand the full path part. Do you mean the physical path?

**Comment by milaniliev** [comment]
Well, in either case this is the physical path on the host. The colon syntax is the path relative to the user's (`git`'s) home dir, and the second path is the absolute path.

**Comment by madarche** [comment]
The git regression is still present in npm@2.11.0

**Comment by madarche** [comment]
The git regression is still present in npm@2.11.1

**Comment by iarna** [comment]
The fact that this ticket is still open should hopefully make it unsurprising that this issue still exists. 😉

For the record, this is being caused by this code:

https://github.com/npm/normalize-git-url/blob/master/normalize-git-url.js#L15

And the module and thus this issue were introduced here:

https://github.com/npm/npm/commit/1c48d08dea31a11ac11a285cac598a482481cade

It transforms the bogus not-urls that git's ssh support accepts into a valid url. Unfortunately when it does, it changes the meaning. The not-urls refer to relative paths, and the true-urls can't do that. It needs to leave that part alone.

**Comment by thefourtheye** [comment]
@iarna what do you mean by not-urls and true-urls?

**Comment by iarna** [comment]
@thefourtheye 
scp style: `user@host:relative/path/to/thing`
True urls are: `git+ssh://user@host:port/absolute/path/to/thing`
Not-urls by contrast are a mix of the two above: `git+ssh://user@host:relative/path/to/thing`

Unfortunately due to constraints of git and URLs, the not-urls are unavoidable if you want relative pathing. 😕

**Comment by thefourtheye** [comment]
Actually when I tested, the scp style works with local relative path and absolute paths. Have to try it with github URLs.

**Comment by fredrikekelund** [comment]
To complicate matters even further: the relative path can contain colons, too. For example, my git server is hosted on a Windows machine, which means the git URL's in my package.json look something like `git+ssh://git@foo.bar.com:C:/Git/reponame.git`

**Comment by zkat** [comment]
Note: `git+ssh://git@foo.com:this/is/relative` is an invalid URI, afaict. At the _very least_, trying to clone a github repo with that syntax fails. Like @iarna said, `scp` syntax is a special case where the specifier really isn't a URI. It's unclear whether specific git servers would know what to do with something like that, but I wouldn't hold my breath about it. You most likely just want to just omit the protocol -- `git@foo.bar.com:relative/path/to/repo.git` *is* a valid `scp`-style identifier, and git will almost certainly be able to understand that. This won't work quite yet, though, because there's an actual bug here where we're breaking valid `scp` identifiers (by putzing with the path and shoving an `ssh://` protocol in front of it.

We should also not normalize _at all_ for non-known-hosts that we're not sure actually have the same data behind multiple resources, because we don't actually know that they're actually putting the same repo behind two different URIs with similar-looking paths.

tl;dr there's three issues:

* `git+ssh://git@foo.bar.com:relative/path/here.git` is invalid. If you want `scp`-style syntax, get rid of the `git+ssh`. That's on you.
* We are normalizing perfectly valid URIs wrong (`git@github.com:foo/bar.git` is getting an `ssh://` added to it and the path is being converted to `/foo/bar.git`)
* We are normalizing URIs that we don't know actually represent the same resource under multiple names. Github and other hosts are regular enough that we know `https://github.com/foo/bar` and `git@github.com:foo/bar` are two ways of accessing the exact same repo, but that's entirely up to the hosts -- we should not normalize *at all* for anything other than these.

I'm tackling the last two now. Also note that this is related to #3224 and a variety of other issues we've seen in the past few years. With any luck, we can wrap this up once and for all :grin:.

**Comment by zkat** [comment]
For reference while I jump on this, here are the docs for what git _actually_ accepts, url-wise:

https://www.kernel.org/pub/software/scm/git/docs/git-fetch.html#URLS

**Comment by madarche** [comment]
@zkat my concern is just to be able to install NPM modules from local-and-private Git repositories accessible only through SSH. I've just been using the [Git URLs specified in the NPM doc](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies). I'm not pushing for a particular syntax and would be most happy to use any valid URIs and provide you with all the tests you need, etc.

With npm@2.11.3 I've tried the `git@foo.bar.com:relative/path/here.git` and `git@foo.bar.com:/absolute/path/here.git` syntaxes (that is without the `git+ssh://` prefix) with no success.

**Comment by milaniliev** [comment]
@madarche I've had success with `git+ssh://git@foo.com/absolute/path.git`. It's not ideal (I'd rather not hard-code the absolute path to my git directory on every machine) but it worked, at least recently.

**Comment by madarche** [comment]
Thanks a lot @milaniliev!

I confirm that `git+ssh://git@foo.com/absolute/path.git` works fine both with npm@2.11.3 (npm@latest at the moment) and npm@2.10.1 bundled with the latest Node.js stable (v0.12.4). And it's great news for me :-) Thanks to all!

@iarna you did give the solution too, but your formulation was not as clear as milaniliev's :-)

A minimal fix for this is to update the [Git URLs specified in the NPM doc](https://docs.npmjs.com/files/package.json#git-urls-as-dependencies) to specify `git+ssh://git@foo.com/absolute/path.git`.

**Comment by svrcekmichal** [comment]
Is there some working solution with bitbucket? I have tried 

```
npm install git+ssh://git@bitbucket.org:username/repository.git
npm install git+ssh://git@bitbucket.org/username/repository.git
```

with no success.

**Comment by othiym23** [comment]
This was fixed in npm itself in 94ca4a711619ba8e40ce3d20bc42b13cdb7611b7, which will be part of the next `npm@2` release (probably `npm@2.12.1`).

**Comment by madarche** [comment]
I confirm that scp-style Git URLs work again in npm@2.12.1. Great! And thanks!

@zkat your work is much much much appreciated. And thanks to you all!

PS: Without wanting to be disrespectful I have tested with [newly announced npm@3](http://blog.npmjs.org/post/122450408965/npm-weekly-20-npm-3-is-here-ish#_=_) and it doesn't work. I guess @zkat's work needs to be merged/ported?

**Comment by zkat** [comment]
This is on `npm@3` branch as of f2bca59e, and should be published with the next release. Hope this is the end of this! :)
